### ISSUE TITLE
supershell breaks interactive programs

### ISSUE THREAD

**Issue by smarter** [sbt/sbt#5352]
```shell
git clone https://github.com/smarter/dotty -b upgrade/sbt-1.3.6.bak 
cd dotty
sbt -Dsbt.supershell=false repl # this works fine
sbt repl # this is unusable
```

**Comment by eatkins** [comment]
The `console` task (along with some others) is specifically blacklisted so that supershell rendering is disabled while that task is active. The question is what is the best way to blacklist user or plugin defined interactive tasks? I could imagine adding a key like `val interactiveTasks = taskKey[Seq[String]]("Tasks that read user input during which progress line rendering should be disabled")` and then plugin could add something like `ThisBuild / interactiveTasks += "repl"`.

**Comment by smarter** [comment]
`repl` just calls `runMain` with custom arguments, I would expect `runMain` to do the right thing to run my cli application without having to customize anything in my build.

**Comment by eatkins** [comment]
Aah, cool. That should be fixed in 1.4.0 for sure. Also I improved supershell rendering so that it should actually work in many case with interactive tasks. I just disabled the supershell blacklisting for console on my local build and recorded this: https://asciinema.org/a/NWOhWbyQfAvDR8Pjph9XeyrzZ

**Comment by eatkins** [comment]
Also, `runMain` isn't blacklisted but `run` is. That would be an easy 1.3.7 fix.

**Comment by smarter** [comment]
What was changed in 1.4.0 ?

**Comment by eatkins** [comment]
There is no short answer to that question. I am in the middle of overhauling how sbt manages user input from both the console (input received from stdin on the sbt server process) and from the client. There is an open [PR](https://github.com/sbt/sbt/pull/5319) that starts this work but I have a private branch with much more extensive changes. The hope is that for sbt 1.4.0, by default we spin up an sbt daemon server and connect to it with the thin client. If you watched the video I posted above, you may have noticed that sbt started up almost instantaneously. That's because I built graalvm image of the sbt client that connects to the running server or starts one up if it isn't running already and then connects to it. A lot of changes are necessary to make all of the features, including supershell, work with both the thin client and when the user is directly connected to the server (the normal interactive mode today).

**Comment by eatkins** [comment]
But the reason supershell behaves differently is that in my new branch, the server keeps track of all output that is printed to a terminal session. Given that it knows the dimensions of the terminal, it can restore the cursor position correctly after re-printing the status lines. There is also additional synchronization to ensure that printing to stdout in tasks is never interleaved with printing out the progress lines.

**Comment by dwijnand** [comment]
> then plugin could add something like `ThisBuild / interactiveTasks += "repl"`

(This suggestion is old, but I thought I would still share: this is a good, rare example of where `Global` should be used over `ThisBuild`: super-shell is an sbt-wide aspect, like the global cancellation thing.  So `Global / interactiveTasks += "repl"`.)

**Comment by jamesward** [comment]
Just ran into this as well.  With supershell on, the stdout gets clobbered when run with `runMain` on a simple app:
```
object Foo extends App {
  Iterator.continually(StdIn.readLine("> ")).takeWhile(_.nonEmpty).foreach(println)
}
```

**Comment by eed3si9n** [comment]
This was fixed in https://github.com/sbt/sbt/pull/5353 by @eatkins and I just backported it so hopefully it will be fixed in the next 1.3.x.

**Comment by jamesward** [comment]
Confirmed fixed in 1.3.7 - thanks!
